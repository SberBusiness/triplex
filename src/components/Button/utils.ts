import React from 'react';

/**
 * Не происходит фокус при клике на button в сафари и в мозиле на маках.
 * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
 *
 * Данная функция должна вызываться только синхронно. (Не будет работать в колбэках setState)
 * https://stackoverflow.com/questions/42089795/reactjs-cant-set-state-from-an-event-with-event-persist
 */
export const focusButton = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.currentTarget?.focus();
};
