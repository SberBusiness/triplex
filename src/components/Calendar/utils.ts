import moment from 'moment';
import {headerDateFormat} from '@sberbusiness/triplex/consts/DateConst';
import {ECalendarViewMode} from '@sberbusiness/triplex/components/Calendar/enums';
import {TPickedDate, TPickedDateProp} from '@sberbusiness/triplex/components/Calendar/types';
import {ICalendarRangeProps, TCalendarProps} from '@sberbusiness/triplex/components/Calendar/Calendar';

/**
 * Приведение даты к типу Moment.
 * @param value Значение.
 * @param format Формат для значения.
 */
export function parsePickedDate(value: TPickedDateProp | undefined, format: string): TPickedDate {
    if (!value) {
        return null;
    } else if (typeof value === 'string') {
        return moment(value, format);
    } else {
        return value;
    }
}

/**
 * Получить актуальный заголовок.
 * @param {TPickedDate} pickedDate Выбранная дата в формате Moment.
 * @param {Moment} [date] Дата для построения заголовка.
 */
export function getHeader(pickedDate: TPickedDate, date?: moment.Moment): string {
    let currentDate;
    if (date && date.isValid()) {
        currentDate = date;
    } else if (pickedDate && pickedDate.isValid()) {
        currentDate = pickedDate;
    } else {
        currentDate = moment();
    }

    return currentDate.format(headerDateFormat);
}

/**
 * Получить текст заголовка из текущей даты.
 * @param date Отображаемая дата.
 * @param tab Текущая вкладка.
 */
export function formatDate(date: TPickedDate, tab: ECalendarViewMode): string | null {
    const checkedDate = date ? date : moment();

    const yearFrom = checkedDate.clone().add(-5, 'y').format('YYYY');
    const yearTo = checkedDate.clone().add(6, 'y').format('YYYY');

    switch (tab) {
        case ECalendarViewMode.YEARS:
            return `${yearFrom} - ${yearTo}`;
        case ECalendarViewMode.MONTHS:
            return checkedDate.clone().format('YYYY');
        case ECalendarViewMode.DAYS:
            return getHeader(checkedDate);
        default:
            return null;
    }
}

/** Type guard для проверки свойств календаря на выбор периода. */
export function isCalendarRange(props: TCalendarProps): props is ICalendarRangeProps {
    return Array.isArray(props.pickedDate);
}
