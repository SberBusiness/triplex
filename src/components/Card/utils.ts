import {ECardRoundingSize} from '@sberbusiness/triplex/components/Card/enums';

/** Соответствие типа скругления карточки стилевому классу. */
export const mapCardRoundingSizeToCssClass = {
    [ECardRoundingSize.MD]: 'cssClass[roundingMD]',
    [ECardRoundingSize.SM]: 'cssClass[roundingSM]',
};
