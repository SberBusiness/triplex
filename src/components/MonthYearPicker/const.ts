/** Формат даты, отображаемый при вооде. */
export const inputMonthYearFormat = 'MMM YYYY';

/** Стандартное значение плейсхолдера. */
export const defaultMonthYearPlaceholder = 'мм.гггг';
