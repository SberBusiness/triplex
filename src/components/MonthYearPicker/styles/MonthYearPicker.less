@import 'src/styles/style';

.monthYearPicker {
    position: relative;
    width: @picker-width;
}
