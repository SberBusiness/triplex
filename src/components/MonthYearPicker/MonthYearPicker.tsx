import React, {useEffect, useState} from 'react';
import moment from 'moment';
import {DatePickerExtended, IDatePickerExtendedProps} from '@sberbusiness/triplex/components/DatePickerExtended/DatePickerExtended';
import {dateFormatYYYYMMDD, globalLimitRange} from '@sberbusiness/triplex/consts/DateConst';
import {MonthYearPickerUtils} from '@sberbusiness/triplex/components/MonthYearPicker/utils';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {MonthYearPickerTarget} from '@sberbusiness/triplex/components/MonthYearPicker/components/MonthYearPickerTarget';
import {MonthYearPickerDropdownHeaderTarget} from '@sberbusiness/triplex/components/MonthYearPicker/components/MonthYearPickerDropdownHeaderTarget';
import {ECalendarPickType} from '@sberbusiness/triplex/components/Calendar/enums';

/** Свойства компонента MonthYearPicker. */
export interface IMonthYearPickerProps
    extends Omit<IDatePickerExtendedProps, 'pickedDate' | 'onDateChange' | 'renderTarget' | 'renderDropdownHeaderTarget'> {
    /** Значение даты. */
    value: string;
    /** Текст подсказки. */
    placeholder?: string;
    /** Неактивное состояние. */
    disabled?: boolean;
    /** Состояние ошибки. */
    error?: boolean;
    /** Функция, вызывающаяся при изменении значения. */
    onChange: (value: string) => void;
}

/** Поле для выбора месяца */
export const MonthYearPicker = React.forwardRef<HTMLDivElement, IMonthYearPickerProps>((props, ref) => {
    const {
        className,
        value,
        placeholder,
        'aria-label': ariaLabel,
        'aria-labelledby': ariaLabelledby,
        format = dateFormatYYYYMMDD,
        limitRange = globalLimitRange,
        disabled,
        error,
        onChange,
        ...rest
    } = props;
    const [pickerValues, setPickerValues] = useState(MonthYearPickerUtils.getPickerValues(value, format, limitRange));

    useEffect(() => {
        const newPickerValues = MonthYearPickerUtils.getPickerValues(value, format, limitRange);

        if (newPickerValues.inputString !== pickerValues.inputString) {
            setPickerValues(newPickerValues);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [value, format, limitRange]);

    /** Рендер-функция управляющего элемента. */
    const renderTarget = () => (
        <MonthYearPickerTarget>
            <MonthYearPickerTarget.Input
                value={pickerValues.inputString}
                placeholder={placeholder}
                aria-label={ariaLabel}
                aria-labelledby={ariaLabelledby}
                disabled={disabled}
                error={error}
            />
            <MonthYearPickerTarget.Button disabled={disabled} />
        </MonthYearPickerTarget>
    );

    /** Рендер-функция управляющего элемента в заголовке DropdownMobile. */
    const renderDropdownHeaderTarget = () => {
        return <MonthYearPickerDropdownHeaderTarget value={pickerValues.inputString} placeholder={placeholder} />;
    };

    /** Обработчик выбора даты. */
    const handleDateChange = (date: moment.Moment) => {
        onChange(date.format(format));
    };

    return (
        <DatePickerExtended
            className={classnames('cssClass[monthYearPicker]', className)}
            renderTarget={renderTarget}
            renderDropdownHeaderTarget={renderDropdownHeaderTarget}
            pickedDate={pickerValues.calendarDate}
            pickType={ECalendarPickType.monthYearPick}
            limitRange={limitRange}
            onDateChange={handleDateChange}
            {...rest}
            ref={ref}
        />
    );
});

MonthYearPicker.displayName = 'MonthYearPicker';
