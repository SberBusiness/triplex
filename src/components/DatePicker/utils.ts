import moment from 'moment';
import {TPickedDate} from '@sberbusiness/triplex/components/Calendar/types';
import {IDateLimitRange} from '@sberbusiness/triplex/types/DateTypes';
import {inputDateFormat} from '@sberbusiness/triplex/components/DatePicker/const';

/** Значения DatePicker. */
interface IDatePickerValues {
    inputString: string;
    calendarDate: TPickedDate;
}

/** Утилиты для компонента DatePicker. */
export const DatePickerUtils = {
    /** Получить значения пикера. */
    getPickerValues: (value: string, format: string, limitRange: IDateLimitRange, disabledDays?: string[]): IDatePickerValues => {
        if (value.length === 0) {
            return {calendarDate: null, inputString: ''};
        }

        const date = DatePickerUtils.getCalendarDate(value, format, limitRange, disabledDays);

        if (date === null) {
            return {calendarDate: null, inputString: ''};
        }

        return {calendarDate: date, inputString: date.format(inputDateFormat)};
    },

    /** Получить календарную дату. */
    getCalendarDate: (value: string, format: string, limitRange: IDateLimitRange, disabledDays?: string[]) => {
        const date = moment(value, format, true);

        if (
            !date.isValid() ||
            date.isBefore(limitRange.dateFrom, 'day') ||
            date.isAfter(limitRange.dateTo, 'day') ||
            disabledDays?.includes(value)
        ) {
            return null;
        }

        return date;
    },
};
