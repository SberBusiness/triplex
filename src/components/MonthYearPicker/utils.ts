import moment from 'moment';
import {TPickedDate} from '@sberbusiness/triplex/components/Calendar/types';
import {IDateLimitRange} from '@sberbusiness/triplex/types/DateTypes';
import {inputMonthYearFormat} from '@sberbusiness/triplex/components/MonthYearPicker/const';

/** Значения MonthYearPicker. */
interface IMonthYearPickerValues {
    inputString: string;
    calendarDate: TPickedDate;
}

/** Утилиты для компонента MonthYearPicker. */
export const MonthYearPickerUtils = {
    /** Получить значения пикера. */
    getPickerValues: (value: string, format: string, limitRange: IDateLimitRange): IMonthYearPickerValues => {
        if (value.length === 0) {
            return {calendarDate: null, inputString: ''};
        }

        const date = MonthYearPickerUtils.getCalendarDate(value, format, limitRange);

        if (date === null) {
            return {calendarDate: null, inputString: ''};
        }

        return {calendarDate: date, inputString: date.format(inputMonthYearFormat)};
    },

    /** Получить календарную дату. */
    getCalendarDate: (value: string, format: string, limitRange: IDateLimitRange) => {
        const date = moment(value, format, true);

        if (!date.isValid() || !MonthYearPickerUtils.isAvailableDate(date, value, limitRange)) {
            return null;
        }

        return date;
    },

    /** Является ли дата доступной. */
    isAvailableDate: (date: moment.Moment, value: string, limitRange: IDateLimitRange) => {
        return !(date.isBefore(limitRange.dateFrom, 'day') || date.isAfter(limitRange.dateTo, 'day'));
    },
};
