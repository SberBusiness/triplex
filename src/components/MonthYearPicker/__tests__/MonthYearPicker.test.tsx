import React from 'react';
import moment from 'moment';
import renderer from 'react-test-renderer';
import {MonthYearPicker} from '@sberbusiness/triplex/components/MonthYearPicker/MonthYearPicker';
import {getFormattedDate} from '@sberbusiness/triplex/utils/dateUtils';

jest.mock('@sberbusiness/icons/SpinnersmallAniIcon20', () => ({
    SpinnersmallAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/SpinnersmallwhiteAniIcon20', () => ({
    SpinnersmallwhiteAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/CalendarSrvIcon20', () => ({
    CalendarSrvIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretleftSrvxIcon24', () => ({
    CaretleftSrvxIcon24: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretrightSrvxIcon24', () => ({
    CaretrightSrvxIcon24: 'svg',
}));

jest.mock('@sberbusiness/icons/ClosenotificationSrvxIcon16', () => ({
    ClosenotificationSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/HeaderkebabSrvxIcon16', () => ({
    HeaderkebabSrvxIcon16: 'svg',
}));

describe('MonthYearPicker', () => {
    beforeEach(() => {
        allure.feature('DatePicker');
    });

    it('renders correctly', () => {
        const handleChange = jest.fn();
        const value = getFormattedDate(moment('1970-01-01'));
        const tree = renderer.create(<MonthYearPicker value={value} onChange={handleChange} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
