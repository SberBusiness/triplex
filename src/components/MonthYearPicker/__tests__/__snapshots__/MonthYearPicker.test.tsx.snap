// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthYearPicker renders correctly 1`] = `
<div
  className="cssClass[monthYearPicker]"
  data-tx="10.0.0"
  onKeyDown={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="cssClass[monthYearPickerTarget]"
  >
    <input
      className="cssClass[input]"
      data-tx="10.0.0"
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="мм.гггг"
      readOnly={true}
      role="button"
      value="Jan 1970"
    />
    <button
      className="cssClass[buttonIcon] cssClass[squircle] hoverable cssClass[datePickerTargetButton]"
      data-tx="10.0.0"
      onClick={[Function]}
      role="presentation"
      tabIndex={-1}
      type="button"
    >
      <svg />
    </button>
  </div>
</div>
`;
